### PR TITLE
fetch: fix FetchTasklet UAF + leak when worker terminates mid-request

### PR DESCRIPTION
**Pull request by Jarred-Sumner**
## What

Fixes a use-after-free + a leak when `Worker.terminate()` runs while a `fetch()` (with or without an `AbortSignal`) is still in flight.

## Root cause

`FetchTasklet.javascript_vm` is a raw `*VirtualMachine` pointing at the worker's VM, which is allocated in the worker's `MimallocArena`. `WebWorker.exitAndDeinit` destroys that arena while the HTTP thread may still be holding the tasklet, so any HTTP-thread read of `task.javascript_vm` afterwards is UAF. The existing code did exactly that in three places:

- `callback()` checked `task.javascript_vm.isShuttingDown()` and called `task.javascript_vm.eventLoop().enqueueTaskConcurrent(...)`
- `derefFromThread()` did the same, and additionally ran **full** `deinit()` on the HTTP thread when it thought the VM was shutting down — calling `clearAbortSignal()` (non-atomic `RefCounted<AbortSignal>::deref`) and `jsc.Strong`/`jsc.Weak` deinits, none of which are thread-safe
- `onWriteRequestDataDrain()` did the same check + enqueue

Separately, `callback()`'s shutdown early-return skipped enqueuing `onProgressUpdate` for the final chunk but never dropped the matching ref, so the tasklet, its `AbortSignal` ref + native callback, and several `jsc.Strong`s leaked.

## Fix

- Add `VirtualMachine.Handle` — `{ScriptExecutionContextIdentifier, ?*VirtualMachine}`. `get()` re-validates against the global `ScriptExecutionContext` map (which `getScriptExecutionContext()` already locks) and returns null once the context is unregistered (the map entry is removed in `~GlobalObject()`). Once null is observed it sticks, so subsequent calls skip the lookup.
- Replace `FetchTasklet.javascript_vm: *VirtualMachine` with `vm_handle: VirtualMachine.Handle`. HTTP-thread call sites (`callback`, `derefFromThread`, `onWriteRequestDataDrain`) use `vm_handle.get() orelse <thread-safe cleanup>`; JS-thread call sites use `vm_handle.get().?`. There is an inherent TOCTOU between `get()` returning non-null and using the pointer; that window is much narrower than reading the raw pointer directly, and avoids per-tasklet bookkeeping.
- `derefFromThread()` now calls a new `deinitFromThread()` that frees only `bun.default_allocator`-owned memory (buffers, headers, metadata, the http allocation, the tasklet itself) and intentionally leaks `AbortSignal`/`jsc.Strong`/`jsc.Weak`/`native_response`. Leaking at VM teardown is correct; racing is not.
- `callback()`'s dead-VM/shutdown branch now resets `has_schedule_callback` and drops the ref `onProgressUpdate` would have dropped, so the deferred `derefFromThread` reaches zero and runs thread-safe cleanup instead of leaking the whole tasklet.

## Test

`test/js/web/fetch/fetch-worker-terminate-shutdown.test.ts` spawns a subprocess that loops 20×: start a Worker that `fetch()`es a slow streamed body with an `AbortSignal`, wait for it to confirm the request is in flight, then `terminate()` it. Asserts the subprocess prints `PASS 20`, has empty stderr, and exits 0.

- 5/5 pass on this branch (`bun bd test`)
- 3/3 fail on system bun (segfault / ASAN use-after-poison on the HTTP thread)
- `fetch.test.ts` (341 tests) still passes

### CODE REVIEW

**Comment by robobun** [comment]
<div><sup>Updated 10:24 PM PT - Apr 14th, 2026</sup></div>

@Jarred-Sumner, your commit 1e6da2d03857fb6d89c2f37b3c4021568eb6ddb2 is building: [`#45767`](https://buildkite.com/bun/bun/builds/45767)

File: src/bun.js/VirtualMachine.zig [hunk 1]
```diff
@@ -271,6 +271,34 @@ pub fn isShuttingDown(this: *const VirtualMachine) bool {
...
+        return self.#vm;
+    }
+};
+
+pub fn getHandle(self: *VirtualMachine) Handle {
+    return .{
+        .#id = self.global.scriptExecutionContextIdentifier(),
+        .#vm = self,
+    };
+}
```
[the remark this answers is not in the document]
Comment (Jarred-Sumner) [reply src/bun.js/VirtualMachine.zig:300]: Acknowledged and intentional — see the doc comment on `Handle` (lines 274–283). The TOCTOU window is strictly narrower than the prior raw `*VirtualMachine` deref it replaces; closing it fully would require ref-counting the worker arena, which is out of scope here.

File: src/bun.js/webcore/fetch/FetchTasklet.zig [hunk 2]
```diff
@@ -269,6 +280,53 @@ pub const FetchTasklet = struct {
...
+        if (this.metadata) |*md| {
+            md.deinit(allocator);
+            this.metadata = null;
+        }
+        this.response_buffer.deinit();
+        this.scheduled_response_buffer.deinit();
+        if (this.request_body_streaming_buffer) |buffer| {
+            this.request_body_streaming_buffer = null;
+            buffer.clearDrainCallback();
+            buffer.deref();
```
[the remark this answers is not in the document]
Comment (Jarred-Sumner) [reply src/bun.js/webcore/fetch/FetchTasklet.zig:320]: Fixed in d6807f460f. Kept only the `.Sendfile` arm — `fd.close()` is a pure syscall and FDs are finite, so worth closing even at shutdown.

Dropped `.AnyBlob`: `AnyBlob.detach()` can reach `WTFStringImpl.deref()` (non-atomic refcount, Blob.zig:4537), so it is not safe from the HTTP thread. `.ReadableStream` stays leaked (jsc.Strong). A leaked Blob.Store at VM teardown is acceptable; a leaked FD is not.

[the remark this answers is not in the document]
Comment (Jarred-Sumner) [reply src/bun.js/webcore/fetch/FetchTasklet.zig:320]: Updated in 2611004c31 — `.AnyBlob` is now detached too. I was wrong about `WTFStringImpl`: `WTF::StringImpl::m_refCount` is `std::atomic<uint32_t>` (StringImpl.h:163), so all three `AnyBlob` variants are thread-safe to detach (`Blob.Store` atomic, `WTFStringImpl` atomic, `InternalBlob` just `bun.default_allocator.free`). Only `.ReadableStream` (jsc.Strong) is still intentionally leaked.

File: src/bun.js/webcore/fetch/FetchTasklet.zig [hunk 3]
```diff
@@ -1440,9 +1509,28 @@ pub const FetchTasklet = struct {
...
+            // Reset the flag we just set so a later final-chunk callback isn't
+            // gated; for the final chunk, drop the ref onProgressUpdate would
+            // have dropped so the deferred derefFromThread() above reaches
+            // zero (after mutex.unlock) and runs thread-safe cleanup.
+            task.has_schedule_callback.store(false, .monotonic);
+            if (is_done) {
+                const old = task.ref_count.fetchSub(1, .monotonic);
+                bun.debugAssert(old > 1);
+            }
+            return;
```
[the remark this answers is not in the document]
Comment (Jarred-Sumner) [reply src/bun.js/webcore/fetch/FetchTasklet.zig:1523]: Correct analysis — narrow leak (not crash) in the streaming-upload + worker-terminate window. Out of scope for this PR, which already turns the prior UAF into at-worst-a-leak. Tracking as a follow-up; the likely fix is `ref_count.store(1, .monotonic)` when `is_done && vm == null` since no other owner remains.
